Performance: Cache django-guardian permissions when counting documents

Merinorus opened the pull request:
<!--
Note: All PRs with code changes should be targeted to the `dev` branch, pure documentation changes can target `main`
-->

## Proposed change

Removes N+1 queries in tag, correspondent, storage path, custom field, and document type list views.
This reduces SQL queries from 160 to 9, and speeds up (~2-5x) the load time and search time of the following pages:
- Correspondents
- Tags
- Document Types
- Storage Paths
- Custom Fields


## TODO

- [x] Test a view
- [x] rename `users_with_view` -> `user_view_perms` etc.

## Feedback welcome

Initially, I wanted to solve this by prefetching permissions with Django-guardian, as they suggest in their documentation: https://django-guardian.readthedocs.io/en/latest/userguide/performance/

Unfortunately, I didn't manage to do it, so I ended up caching them myself in the view and retrieving them in the serializer.

This is a cool improvement on slow machines, but the code is too complex and can probably be simplified. Maybe it has to be rewritten completely. Feedback and suggestions are highly appreciated!

## Type of change

<!--
What type of change does your PR introduce to Paperless-ngx?
NOTE: Please check only one box!
-->

- [x] Bug fix: non-breaking change which fixes an issue.
- [ ] New feature / Enhancement: non-breaking change which adds functionality. _Please read the important note above._
- [ ] Breaking change: fix or feature that would cause existing functionality to not work as expected.
- [ ] Documentation only.
- [ ] Other. Please explain:

## Checklist:

<!--
NOTE: PRs that do not address the following will not be merged, please do not skip any relevant items.
-->

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [ ] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [ ] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

Merinorus committed:
Performance: Cache django-guardian permissions when counting documents

Fixes N+1 queries in tag, correspondent, storage path, custom field,
and document type list views.
Reduces SQL queries from 160 to 9.

stumpylog commented:
I'm concerned that caching won't reflect changes to a user's permissions of groups.  When does it get cleared/updated?  Was the problem with the `ObjectPermissionChecker` that we also have the owner idea?

Merinorus committed:
Revert "Performance: Cache django-guardian permissions when counting documents"

This reverts commit da2a92eaccf136592c106ba5fe6eb1237210c425.

Merinorus committed:
Preload django-guardian permissions for lists

Avoid N+1 queries when loading list of Correspondents, Tags, Document Types of Storage Paths.

Merinorus committed:
One request for a list of permission instead of one request per permission

Merinorus committed:
Query only required fields for permissions

Merinorus committed:
Query permissions for current page only

Merinorus committed:
Add test & Code cleaning

shamoon commented:
Oh nevermind, I see that was just an internal name

Merinorus committed:
Code simplification

Merinorus committed:
Test that no N+1 queries are made

Merinorus commented:
Ok, I think it's easier to read now. I also added a test to check N+1 queries are not made anymore.
As expected, this test fails on the dev branch:
``` AssertionError: Possible N+1 queries detected: 23 queries for 2 tags, but 158 queries for 50 tags```

Merinorus committed:
fix docstring

Merinorus commented on src/documents/serialisers.py:
```diff
@@ -251,6 +252,35 @@ def __init__(self, *args, **kwargs):
+        """
+        Get the given permissions from context or from django-guardian.
+        Parameters:
+        - codename: e.g. 'view_tag' or 'change_tag'
```
Yes, good catch, thank you. I fixed the docstring.
